MAINT Remove loguniform fix, use scipy.stats instead, bump scipy version and Ubuntu version in CI

betatim opened the pull request:
<!--
Thanks for contributing a pull request! Please ensure you have taken a look at
the contribution guidelines: https://github.com/scikit-learn/scikit-learn/blob/main/CONTRIBUTING.md
-->

#### Reference Issues/PRs
Continues #23730 and closes #23699 closes


#### What does this implement/fix? Explain your changes.
This removes `sklearn.fixes.loguniform` and uses `scipy.stats.loguniform` instead. It also updates the docs where they refer to `loguniform` being special because it is provided by `sklearn.fixes`.

#### Any other comments?
For now bumping scipy to v1.5.0 to see what the CI says. Will read the discussions on which version exactly to bump to.

PR is on hold because of https://github.com/scikit-learn/scikit-learn/issues/24401#issuecomment-1277549665. So maybe this PR can be useful at a later point.

betatim committed:
Remove loguniform fix, use scipy.stats instead

betatim committed:
Bump scipy to v1.5.0

ogrisel commented:
As discussed in https://github.com/scikit-learn/scikit-learn/issues/24401#issuecomment-1277549665 we want to keep scipy 1.3 / ubuntu 20.04 compat in scikit-learn 1.2 so this PR can only be merged to `main` after `1.2.X` is branched off.

ogrisel commented:
@betatim maybe now is the time to revive this :)

betatim committed:
Merge branch 'main' into remove-loguniform-fix

betatim committed:
Remvove obsolete loguniform fix

betatim committed:
Updated environment and lock files for scipy 1.5

betatim commented:
Let's see what the bots have to say.

ogrisel committed:
Merge branch 'main' into remove-loguniform-fix

ogrisel reviewed (commented):
I think the minimum version for the dependencies documented in the `README.rst` file have to be updated manually.

Also there is a paragraph in `install.rst` that needs to be updated (or removed if no longer applicable):

```
.. note::

   For installing on PyPy, PyPy3-v5.10+, Numpy 1.14.0+, and scipy 1.1.0+
   are required.
```

ogrisel reviewed (approved):
LGTM, assuming the above comments are addressed and the CI is green (on a commit with a message that enabled both a full doc build and a pypy build).

lorentzenchr reviewed (approved):
LGTM. Once Olivier‘s comments are addressed.

betatim commented:
The scipy version also appears in the `build-system` section of `pyproject.toml`. I guess it makes sense to also update that right?

lorentzenchr commented:
Could we reduce the occurrences? Updating the min versions should be easy.

betatim committed:
Update scipy version in README and pyproject.toml

betatim committed:
Merge remote-tracking branch 'origin/remove-loguniform-fix' into remove-loguniform-fix

betatim commented:
> Could we reduce the occurrences?

I don't think we can template `README.md` or `pyproject.toml`

thomasjpfan commented:
> Could we reduce the occurrences? Updating the min versions should be easy.

The source of truth for minimal versions is in https://github.com/scikit-learn/scikit-learn/blob/main/sklearn/_min_dependencies.py. We do have the repeat the versions in other places, like in the `README.rst` or `pyproject.toml`, so we do have tests to make sure they are consistent.  (For this PR, the [consistency check fails](https://dev.azure.com/scikit-learn/scikit-learn/_build/results?buildId=51977&view=logs&jobId=dde5042c-7464-5d47-9507-31bdd2ee0a3a&j=dde5042c-7464-5d47-9507-31bdd2ee0a3a&t=4bd2dad8-62b3-5bf9-08a5-a9880c530c94), which is expected.)

That being said, we should have a test that checks that `pyproject.toml` is consistent with `_min_dependencies`. I opened #25610 to add this test.

thomasjpfan commented:
From the failing [ubuntu-atlas](https://dev.azure.com/scikit-learn/scikit-learn/_build/results?buildId=51992&view=logs&j=0a287ed6-22f4-5cb4-88b1-d5fcdc4d8b7e&t=020de790-353b-5fc0-f50d-85f6961db7e7&l=232) job, they ship SciPy 1.3.3. If we still want to test using Ubuntu's SciPy, we would need to update to 22.04LTS: https://packages.ubuntu.com/jammy/python3-scipy

thomasjpfan commented:
As for the note in `install.rst` about PyPy, I would remove it. SciPy does not provide PyPy wheels on [PyPi](https://pypi.org/project/scipy/#files), so it is hard to install with `pip`. Ultimately, the simplest way to use PyPy + scikit-learn is with `conda-forge`:

```bash
mamba create -n pypy pypy scikit-learn
```

For newer versions of PyPy, there are still issues with scikit-learn PyPy migration: https://github.com/conda-forge/scikit-learn-feedstock/pull/213

betatim committed:
Remove PyPy specific install note

betatim commented:
What does the "atlas" refer to in the Ubuntu job? The job runs on [Ubuntu 22.04](https://dev.azure.com/scikit-learn/scikit-learn/_build/results?buildId=51992&view=logs&j=0a287ed6-22f4-5cb4-88b1-d5fcdc4d8b7e&t=6c153403-aea3-40e9-a64b-b9739b499948&l=70) which is the current LTS, with support ending in 2025. So I think installing scikit-learn on 22.04 should be something we test (and support). But maybe "atlas" refers to some particular variant or something?


I've removed the PyPy specific note in `install.rst`.

lorentzenchr commented:
Can we merge?

thomasjpfan commented:
> What does the "atlas" refer to in the Ubuntu job?

Atlas refers to the Linear Algebra library shipped with Ubuntu: https://packages.ubuntu.com/focal/libatlas3-base

betatim commented:
> Can we merge? (after sync with main)

We need to decide if supporting ATLAS (on Ubuntu 20.04) is something scikit-learn cares about. If we bump scipy to 1.5.0 we lose this. This means we can either remove the CI job or increase the version of Ubuntu it uses.

betatim commented:
(aside: why the sync with `main`? My intuition is that it is only needed when there is a conflict, and otherwise we don't need it? As in, I don't quite understand why GitHub even shows the button :-/)

lorentzenchr commented:
The last two issues for increasing scipy where #19705 and #21460.

Ubuntu 18.04 LTS will end [standard support in April 2023](https://wiki.ubuntu.com/Releases). As our 1.3 release will be later (May or June), I guess it's sound to go for Ubuntu 20.04 LTS. If that is still with Atlas (BLAS), then fine.

Maybe we should open a separate issue for that, as this PR is for `loguniform`:smile:

jjerphan commented:
Hi all, I think we already test against Ubuntu 20.04 with ATLAS:

https://github.com/scikit-learn/scikit-learn/blob/18af5508013d8497b0449c059b9a794c9643735a/azure-pipelines.yml#L167-L221

Hence is resolving the `conda` lock-files merge conflicts sufficient for this PR to be merged?

betatim commented:
We test Ubuntu 20.04 with ATLAS in the `scikit-learn.scikit-learn (Linux ubuntu_atlas) ` job, and this is the one that failed. This is why I think [we need to decide if supporting 20.04 + ATLAS is something we care about or not](https://github.com/scikit-learn/scikit-learn/pull/24665#issuecomment-1434263702).

ogrisel commented:
We care about making sure that scikit-learn can work with any commonly shipped implementation of BLAS. But we could consider bumping the dependencies to Ubuntu 22.04 LTS for 1.3.

ogrisel commented:
Could someone re-trigger this PR once with ubuntu 20.04 LTS to see the original error log (it has expired) to confirm the above. And if so, let's try to bump up the Azure build to 22.04 LTS to see if would resolve that last failure as expected.

betatim commented:
I don't know how to bump the build :( But I can tell you that the problem isn't with `loguniform`, it is just a question of which scipy version you install (like you said).

ogrisel commented:
Let's try to bump directly the CI config to 22.04 LTS. You might also need to adapt the matching entry in `build_tools/update_environments_and_lock_files.py` and re-run it.

betatim committed:
Merge branch 'main' into remove-loguniform-fix

betatim committed:
Update Ubuntua ATLAS CI to use 22.04

With Ubuntu 20.04 we can not get a new enough version of scipy, and
22.04 is the latest LTS release.

lorentzenchr commented on azure-pipelines.yml:
```diff
@@ -177,13 +177,26 @@ jobs:
+- template: build_tools/azure/posix.yml
+  parameters:
+    name: Linux
+    vmImage: ubuntu-20.04
```
```suggestion
    vmImage: ubuntu-22.04
```
? I‘m confused.

betatim committed:
Change job name

lorentzenchr reviewed (approved):
Is is now mergable if CI is green?

ogrisel reviewed (approved):
LGTM again.

jjerphan reviewed (approved):
LGTM even if I think having the CI setup be updated in another PR would have been clearer.

I have updated the title to include a `MAINT` prefix. Should we mention something about the CI in the title?
